File: GFG/Minimum sum partition/Minimum sum partition.cpp
class Solution {
public:
   int count_subset(vector<int>a,int sum,int n){
      int dp[n+1][sum+1];
    for(int i=0;i<=n;i++){
        for(int j=0;j<=sum;j++){
            if(i==0)
            dp[i][j]=false;
            if(j==0)
            dp[i][j]=true;
        }
    }            

    for(int i=1;i<=n;i++){
        for(int j=0;j<=sum;j++){
            if(a[i-1]<=j){
                dp[i][j]=dp[i-1][j] || dp[i-1][j-a[i-1]];
            }
            else
            dp[i][j]=dp[i-1][j];
        }
    }

      int minm=INT_MAX;
    for(int i=1;i<sum+1;i++)
    {
        if(dp[n][i]==true)
        {         
            minm=min(minm,abs(i-(sum-i)));
        }
    } 

return minm;
}

    int minimumDifference(vector<int>& nums) {
        if(nums.size()==2)
            return abs(nums[0]-nums[1]);
        else{
        int sum=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
        }
        return count_subset(nums,abs(sum),nums.size());
        }
    }
};

File: GFG/Minimum sum partition/Minimum sum partition.java
class Solution
{

	public int minDifference(int arr[], int n) 
	{ 
	    // Your code goes here
	    Map<String,Integer> map=new HashMap<>();
	    return dfs(arr,n-1,0,0,map);
	} 
	public int dfs(int[]arr,int n,int s1,int s2,Map<String,Integer> map){
	    if(n<0){
	        return Math.abs(s1-s2);
	    }
	    String key=n+","+s1;
	    if(!map.containsKey(key)){
	        int first=dfs(arr,n-1,s1+arr[n],s2,map);
	        int sec=dfs(arr,n-1,s1,s2+arr[n],map);
	        map.put(key,Math.min(first,sec));
	    }
	    return map.get(key);
	}
}
